### PR TITLE
Allow setting `target_family` to multiple values, and implement `target_family="wasm"`

### PR DESCRIPTION
**Pull request by nagisa**
As per the conclusion in [this thread](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Are.20we.20comfortable.20with.20adding.20an.20insta-stable.20cfg.28wasm.29.3F/near/233158441), this implements an ability to specify any number of `target_family` values, allowing for more flexible generic groups, or "families", to be created than just the OS-based unix/windows dichotomy.

cc https://github.com/rust-lang/reference/pull/1006

### CODE REVIEW

**Comment by nagisa** [comment]
Note: this will want at least an FCP for T-lang here or on the reference issue before it is merged.

**Comment by joshtriplett** [comment]
@rfcbot merge

File: compiler/rustc_target/src/spec/mod.rs [hunk 1]
```diff
@@ -1802,7 +1805,7 @@ impl Target {
         key!(exe_suffix);
         key!(staticlib_prefix);
         key!(staticlib_suffix);
-        key!(os_family = "target-family", optional);
+        key!(families = "target-family", list);
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/mod.rs:1808]: You may want to support the old syntax for compatibility, this is an old and potentially popular option.

**Comment by nikomatsakis** [comment]
I'd like to see a better write-up of what is happening here -- perhaps in the form of a PR to some documentation? I can't really understand the expected usage pattern yet.

**Comment by nagisa** [comment]
> perhaps in the form of a PR to some documentation? I can't really understand the expected usage pattern yet.

I… honestly don't know what documentation I'd adjust – the only thing that comes to mind is the book, but adjusting learning materials for something that is only available in nightly seems tad too hasty. I'm also not the one championing for this change – just something I ended up implementing.

All that said, I can quickly (given its 1am here) give a brief overview of the motivations I'm aware of:

It is fairly common to write code that is conditional on the architecture, but does not care particularly much about the pointer width.  Taking x86 as an example, this tends to present itself in the code as `cfg(any(target_arch="x86", target_arch="x86_64"))`. With this proposed change to `target_family` we'd be able to add a "family" that encompasses both `x86` and `x86_64` architectures, reducing the amount of typing necessary down to `cfg(target_family="x86")`.

`wasm` is a much better motivating example in this instance, or so I'm told, because `wasm32` and `wasm64` are only different in their support for indexing memory by 64-bit types (which is in contrast to, say, `x86` and `ARM` which took the opportunity to improve on their instruction set in more involving ways) and so almost every piece of code out there that would be conditional on `wasm` support would have to write `cfg(any(target_arch="wasm32", target_arch="wasm64"))`.

Now,, one might argue that it'd make more sense to instead replace `target_arch="wasm32"` with `target_arch="wasm"` in that case, much like we only have `target_arch="arm"` for the bouquet of 32-bit ARM targets, but the ship appears to have sailed on that one. Expanding the things `target_family` can mean is thus proposed as the backwards compatible solution.

**Comment by nikomatsakis** [comment]
@nagisa I'd be good with writing up a mini-RFC describing the change to the feature :)

**Comment by nikomatsakis** [comment]
So in short, we now have `target_family` as a built-in option, and we have some families defined...is there a table defining which target architectures map to which families?

**Comment by nagisa** [comment]
@nikomatsakis this PR only adds `target_family="wasm"` and other potential groupings were left up to future additions. https://github.com/rust-lang/reference/pull/1006 describes the state after this PR would land as well as the values that would be possible were this PR land.

**Comment by petrochenkov** [comment]
Some rustdoc lints need fixing.

**Pull request merged**